kv/kvserver: TestTenantRateLimiter failed

kv/kvserver.TestTenantRateLimiter [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3471661&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3471661&tab=artifacts#/) on master @ [bfdcfa00836232b910cb0fd109ebedf8913b1466](https://github.com/cockroachdb/cockroach/commits/bfdcfa00836232b910cb0fd109ebedf8913b1466):


```
=== RUN   TestTenantRateLimiter
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestTenantRateLimiter069762263
    test_log_scope.go:80: use -show-logs to present logs inline
    client_tenant_test.go:248: 
        	Error Trace:	client_tenant_test.go:248
        	Error:      	
        	Test:       	TestTenantRateLimiter
    panic.go:613: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestTenantRateLimiter069762263
--- FAIL: TestTenantRateLimiter (23.34s)
```
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestTenantRateLimiter PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-json
</p>
</details>
/cc @cockroachdb/kv ajwerner
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTenantRateLimiter.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver.TestTenantRateLimiter [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3490897&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3490897&tab=artifacts#/) on master @ [85b1e92455eb2795043d71e5766b1b56dab308f2](https://github.com/cockroachdb/cockroach/commits/85b1e92455eb2795043d71e5766b1b56dab308f2):


```
=== RUN   TestTenantRateLimiter
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestTenantRateLimiter750218330
    test_log_scope.go:80: use -show-logs to present logs inline
    client_tenant_test.go:248: 
        	Error Trace:	client_tenant_test.go:248
        	Error:      	
        	Test:       	TestTenantRateLimiter
    panic.go:613: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestTenantRateLimiter750218330
--- FAIL: TestTenantRateLimiter (31.25s)
```
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestTenantRateLimiter PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-json
</p>
</details>
/cc @cockroachdb/kv ajwerner
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTenantRateLimiter.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver.TestTenantRateLimiter [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3493998&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3493998&tab=artifacts#/) on master @ [a0d6d46d3283be18483e3d449adc7d90a02649c1](https://github.com/cockroachdb/cockroach/commits/a0d6d46d3283be18483e3d449adc7d90a02649c1):


```
=== RUN   TestTenantRateLimiter
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestTenantRateLimiter129364174
    test_log_scope.go:80: use -show-logs to present logs inline
    client_tenant_test.go:248: 
        	Error Trace:	client_tenant_test.go:248
        	Error:      	
        	Test:       	TestTenantRateLimiter
    panic.go:613: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestTenantRateLimiter129364174
--- FAIL: TestTenantRateLimiter (31.99s)
```
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestTenantRateLimiter PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-json
</p>
</details>
/cc @cockroachdb/kv ajwerner
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTenantRateLimiter.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/kv/kvserver/client_tenant_test.go
+++ b/pkg/kv/kvserver/client_tenant_test.go
@@ -29,7 +29,6 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/storage/enginepb"
 	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
-	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
 	"github.com/cockroachdb/cockroach/pkg/util/encoding"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
@@ -143,7 +142,6 @@ func TestTenantsStorageMetricsOnSplit(t *testing.T) {
 // and report the correct metrics.
 func TestTenantRateLimiter(t *testing.T) {
 	defer leaktest.AfterTest(t)()
-	skip.WithIssue(t, 70456, "flaky test")
 	defer log.Scope(t).Close(t)
 
 	// This test utilizes manual time to make the rate-limiting calculations more
@@ -163,6 +161,10 @@ func TestTenantRateLimiter(t *testing.T) {
 
 	ctx := context.Background()
 	defer s.Stopper().Stop(ctx)
+
+	// Set a small rate limit so the test doesn't take a long time.
+	runner := sqlutils.MakeSQLRunner(sqlDB)
+	runner.Exec(t, `SET CLUSTER SETTING kv.tenant_rate_limiter.rate_limit = 200`)
 
 	tenantID := serverutils.TestTenantID()
 	codec := keys.MakeSQLCodec(tenantID)
@@ -224,8 +226,7 @@ func TestTenantRateLimiter(t *testing.T) {
 
 	// Create some tooling to read and verify metrics off of the prometheus
 	// endpoint.
-	sqlutils.MakeSQLRunner(sqlDB).Exec(t,
-		`SET CLUSTER SETTING server.child_metrics.enabled = true`)
+	runner.Exec(t, `SET CLUSTER SETTING server.child_metrics.enabled = true`)
 	httpClient, err := s.GetHTTPClient()
 	require.NoError(t, err)
 	getMetrics := func() string {
@@ -247,5 +248,9 @@ func TestTenantRateLimiter(t *testing.T) {
 
 	// Ensure that the metric for the admitted requests reflects the number of
 	// admitted requests.
-	require.Contains(t, getMetrics(), makeMetricStr(int64(tooManyWrites)))
+	// TODO(radu): this is fragile because a background write could sneak in and
+	// the count wouldn't match exactly.
+	m := getMetrics()
+	exp := makeMetricStr(int64(tooManyWrites))
+	require.Contains(t, m, exp, "could not find %s in metrics: \n%s\n", exp, m)
 }

